File: code/matlab/chapter04/igaussrnd.m
function x = igaussrnd(psi, mu, n)
% igaussrnd.m
% Draws from the inverse Gaussian distribution IGAUSS(psi, mu) with
% kernel
%   f(x) \propto x^{-3/2} exp{ -psi (x-mu)^2 / (2 x mu^2) },  x > 0.
% Uses the standard chi-squared-based representation from
% Michael, Schucany, and Haas (1976).
%
% Inputs:
%   psi : positive shape parameter (scalar or n-by-1 vector)
%   mu  : positive mean parameter  (scalar or n-by-1 vector)
%   n   : number of draws (defaults to max(numel(psi), numel(mu)))
%
% Output:
%   x   : n-by-1 vector of draws from IGAUSS(psi, mu)

if nargin < 3
    n = max(numel(psi), numel(mu));
end

% expand scalars to length n; force column vectors
if isscalar(psi), psi = repmat(psi, n, 1); else, psi = psi(:); end
if isscalar(mu),  mu  = repmat(mu,  n, 1); else, mu  = mu(:);  end

if numel(psi) ~= n || numel(mu) ~= n
    error('psi and mu must be scalars or vectors of the same length n.');
end

% step 1: nu0 ~ \chi^2_1
nu0 = randn(n,1).^2;

% step 2: candidate draws
sqrt_term = sqrt(4.*mu.*psi.*nu0 + (mu.^2).*(nu0.^2));
x1 = mu + (mu.^2).*nu0./(2.*psi) - (mu./(2.*psi)).*sqrt_term;
x2 = (mu.^2)./x1;

% step 3: accept/reject switch
p = mu./(mu + x1);
U = rand(n,1) < p;
x = U.*x1 + (~U).*x2;
end

File: code/python/chapter04/igaussrnd.py
# igaussrnd.py
# Draws from the inverse Gaussian distribution IGAUSS(psi, mu) with
# kernel
#   f(x) \propto x^{-3/2} exp{ -psi (x-mu)^2 / (2 x mu^2) },  x > 0.
# Uses the standard chi-squared-based representation from
# Michael, Schucany, and Haas (1976).
#
# Inputs:
#   psi : positive shape parameter (scalar or length-n vector)
#   mu  : positive mean parameter  (scalar or length-n vector)
#   n   : number of draws (defaults to max(psi.size, mu.size))
#
# Output:
#   x   : length-n vector of draws from IGAUSS(psi, mu)

import numpy as np


def igaussrnd(psi, mu, n=None):
    psi = np.atleast_1d(np.asarray(psi, dtype=float)).flatten()
    mu = np.atleast_1d(np.asarray(mu, dtype=float)).flatten()

    if n is None:
        n = max(psi.size, mu.size)

    # expand scalars to length n
    if psi.size == 1:
        psi = np.full(n, psi[0])
    if mu.size == 1:
        mu = np.full(n, mu[0])

    if psi.size != n or mu.size != n:
        raise ValueError('psi and mu must be scalars or vectors '
                         'of the same length n.')

    # step 1: nu0 ~ \chi^2_1
    nu0 = np.random.randn(n)**2

    # step 2: candidate draws
    sqrt_term = np.sqrt(4*mu*psi*nu0 + (mu**2)*(nu0**2))
    x1 = mu + (mu**2)*nu0/(2*psi) - (mu/(2*psi))*sqrt_term
    x2 = (mu**2)/x1

    # step 3: accept/reject switch
    p = mu/(mu + x1)
    U = np.random.rand(n) < p
    x = np.where(U, x1, x2)
    return x
